diff --git a/src/bio.c b/src/bio.c
index 515af26..58a075c 100644
--- a/src/bio.c
+++ b/src/bio.c
@@ -11,6 +11,7 @@ openssl.bio is a help object, it is useful, but rarely use.
 #include <openssl/bn.h>
 #include <openssl/ssl.h>

+#if !defined(OPENSSL_NO_BIO)
 #define MYNAME    "bio"
 #define MYVERSION MYNAME " library for " LUA_VERSION " / Nov 2014 / "\
   "based on OpenSSL " SHLIB_VERSION_NUMBER
@@ -946,3 +947,4 @@ int luaopen_bio(lua_State *L)

   return 1;
 }
+#endif
\ No newline at end of file
diff --git a/src/dh.c b/src/dh.c
index 8375f3e..3a71eef 100644
--- a/src/dh.c
+++ b/src/dh.c
@@ -9,6 +9,7 @@
 #include <openssl/dh.h>
 #include <openssl/engine.h>
 
+#ifndef OPENSSL_NO_DH
 #define MYNAME    "dh"
 #define MYVERSION MYNAME " library for " LUA_VERSION " / Nov 2014 / "\
   "based on OpenSSL " SHLIB_VERSION_NUMBER
@@ -84,3 +85,4 @@ int luaopen_dh(lua_State *L)
   lua_settable(L, -3);
   return 1;
 }
+#endif
diff --git a/src/dsa.c b/src/dsa.c
index 3cb81c9..21f51aa 100644
--- a/src/dsa.c
+++ b/src/dsa.c
@@ -9,6 +9,8 @@
 #include <openssl/dsa.h>
 #include <openssl/engine.h>
 
+#ifndef OPENSSL_NO_DSA
+
 #define MYNAME    "dsa"
 #define MYVERSION MYNAME " library for " LUA_VERSION " / Nov 2014 / "\
   "based on OpenSSL " SHLIB_VERSION_NUMBER
@@ -83,3 +85,4 @@ int luaopen_dsa(lua_State *L)
   lua_settable(L, -3);
   return 1;
 }
+#endif
diff --git a/src/ocsp.c b/src/ocsp.c
index 5ebed3e..0f2ea31 100644
--- a/src/ocsp.c
+++ b/src/ocsp.c
@@ -8,6 +8,7 @@
 #include "private.h"
 #include "openssl/ocsp.h"
 
+#ifndef OPENSSL_NO_OCSP
 #define MYNAME    "ocsp"
 #define MYVERSION MYNAME " library for " LUA_VERSION " / Nov 2014 / "\
   "based on OpenSSL " SHLIB_VERSION_NUMBER
@@ -524,3 +525,4 @@ int luaopen_ocsp(lua_State *L)
 
   return 1;
 }
+#endif
diff --git a/src/openssl.c b/src/openssl.c
index 3ec3bed..595141f 100644
--- a/src/openssl.c
+++ b/src/openssl.c
@@ -526,8 +526,10 @@ LUALIB_API int luaopen_openssl(lua_State*L)
   openssl_register_lhash(L);
   openssl_register_engine(L);

+#ifndef OPENSSL_NO_BIO
   luaopen_bio(L);
   lua_setfield(L, -2, "bio");
+#endif

   luaopen_asn1(L);
   lua_setfield(L, -2, "asn1");
@@ -545,13 +547,15 @@ LUALIB_API int luaopen_openssl(lua_State*L)
   luaopen_pkey(L);
   lua_setfield(L, -2, "pkey");

-#ifdef EVP_PKEY_EC
+#ifndef OPENSSL_NO_EC
   luaopen_ec(L);
   lua_setfield(L, -2, "ec");
 #endif

+#ifndef OPENSSL_NO_X509
   luaopen_x509(L);
   lua_setfield(L, -2, "x509");
+#endif

   luaopen_pkcs7(L);
   lua_setfield(L, -2, "pkcs7");
@@ -559,8 +563,10 @@ LUALIB_API int luaopen_openssl(lua_State*L)
   luaopen_pkcs12(L);
   lua_setfield(L, -2, "pkcs12");

+#ifndef OPENSSL_NO_OCSP
   luaopen_ocsp(L);
   lua_setfield(L, -2, "ocsp");
+#endif

 #ifdef OPENSSL_HAVE_TS
   /* timestamp handling */
@@ -568,8 +574,10 @@ LUALIB_API int luaopen_openssl(lua_State*L)
   lua_setfield(L, -2, "ts");
 #endif

+#ifndef OPENSSL_NO_CMS
   luaopen_cms(L);
   lua_setfield(L, -2, "cms");
+#endif

   luaopen_ssl(L);
   lua_setfield(L, -2, "ssl");
@@ -578,12 +586,18 @@ LUALIB_API int luaopen_openssl(lua_State*L)
   luaopen_bn(L);
   lua_setfield(L, -2, "bn");

+#ifndef OPENSSL_NO_RSA
   luaopen_rsa(L);
   lua_setfield(L, -2, "rsa");
+#endif
+#ifndef OPENSSL_NO_DSA
   luaopen_dsa(L);
   lua_setfield(L, -2, "dsa");
+#endif
+#ifndef OPENSSL_NO_DH
   luaopen_dh(L);
   lua_setfield(L, -2, "dh");
+#endif

 #ifndef OPENSSL_NO_SRP
   luaopen_srp(L);
diff --git a/src/openssl.h b/src/openssl.h
index 33e7255..5c47bdc 100644
--- a/src/openssl.h
+++ b/src/openssl.h
@@ -16,9 +16,8 @@
 #include <assert.h>
 #include <string.h>
 /* OpenSSL includes */
+#include <openssl/opensslconf.h>
 #include <openssl/evp.h>
-#include <openssl/x509.h>
-#include <openssl/x509v3.h>
 #include <openssl/crypto.h>
 #include <openssl/pem.h>
 #include <openssl/err.h>
@@ -113,22 +112,42 @@ LUA_FUNCTION(luaopen_hmac);
 LUA_FUNCTION(luaopen_cipher);
 LUA_FUNCTION(luaopen_bn);
 LUA_FUNCTION(luaopen_pkey);
+#ifndef OPENSSL_NO_X509
+#include <openssl/x509.h>
+#include <openssl/x509v3.h>
 LUA_FUNCTION(luaopen_x509);
+LUA_FUNCTION(luaopen_x509_req);
+LUA_FUNCTION(luaopen_x509_crl);
+#endif
 LUA_FUNCTION(luaopen_pkcs7);
 LUA_FUNCTION(luaopen_pkcs12);
+#ifndef OPENSSL_NO_BIO
 LUA_FUNCTION(luaopen_bio);
+#endif
 LUA_FUNCTION(luaopen_asn1);

+#ifndef OPENSSL_NO_TS
 LUA_FUNCTION(luaopen_ts);
-LUA_FUNCTION(luaopen_x509_req);
-LUA_FUNCTION(luaopen_x509_crl);
+#endif
+#ifndef OPENSSL_NO_OCSP
 LUA_FUNCTION(luaopen_ocsp);
+#endif
+#ifndef OPENSSL_NO_CMS
 LUA_FUNCTION(luaopen_cms);
+#endif
 LUA_FUNCTION(luaopen_ssl);
+#ifndef OPENSSL_NO_EC
 LUA_FUNCTION(luaopen_ec);
+#endif
+#ifndef OPENSSL_NO_RSA
 LUA_FUNCTION(luaopen_rsa);
+#endif
+#ifndef OPENSSL_NO_DSA
 LUA_FUNCTION(luaopen_dsa);
+#endif
+#ifndef OPENSSL_NO_DH
 LUA_FUNCTION(luaopen_dh);
+#endif

 void openssl_add_method_or_alias(const OBJ_NAME *name, void *arg) ;
 void openssl_add_method(const OBJ_NAME *name, void *arg);
@@ -167,7 +186,11 @@ static inline void* openssl_getgroup(lua_State *L, const char* name, int idx)

 int openssl_register_lhash(lua_State* L);
 int openssl_register_engine(lua_State* L);
-
+#if (OPENSSL_VERSION_NUMBER >= 0x10101007L) && !defined(OPENSSL_NO_SM2)
+LUA_FUNCTION(luaopen_sm2);
+#endif
+#ifndef OPENSSL_NO_SRP
 LUA_FUNCTION(luaopen_srp);
-
 #endif
+
+#endif
\ No newline at end of file
diff --git a/src/ots.c b/src/ots.c
index fafd481..c40f23b 100644
--- a/src/ots.c
+++ b/src/ots.c
@@ -7,7 +7,7 @@
 #include "openssl.h"
 #include "private.h"
 #include <stdint.h>
-#if OPENSSL_VERSION_NUMBER >= 0x10000000L
+#if OPENSSL_VERSION_NUMBER >= 0x10000000L && !defined(OPENSSL_NO_TS)
 #include <openssl/ts.h>
 
 #define MYNAME    "ts"
diff --git a/src/rsa.c b/src/rsa.c
index 62deec6..3a57daa 100644
--- a/src/rsa.c
+++ b/src/rsa.c
@@ -9,6 +9,8 @@
 #include <openssl/rsa.h>
 #include <openssl/engine.h>
 
+#ifndef OPENSSL_NO_RSA
+
 #define MYNAME    "rsa"
 #define MYVERSION MYNAME " library for " LUA_VERSION " / Nov 2014 / "\
   "based on OpenSSL " SHLIB_VERSION_NUMBER
@@ -224,3 +226,4 @@ int luaopen_rsa(lua_State *L)
 
   return 1;
 }
+#endif
diff --git a/src/srp.c b/src/srp.c
index 85626b5..2dfcc39 100644
--- a/src/srp.c
+++ b/src/srp.c
@@ -1,6 +1,7 @@
 #include "openssl.h"
 #include "private.h"
 
+#ifndef OPENSSL_NO_SRP
 #include <openssl/srp.h>
 #include <openssl/bn.h>
 
@@ -198,4 +199,4 @@ int luaopen_srp(lua_State *L)
   lua_settable(L, -3);
   return 1;
 }
-
+#endif
diff --git a/src/x509.c b/src/x509.c
index 29213eb..3f06cb5 100644
--- a/src/x509.c
+++ b/src/x509.c
@@ -8,6 +8,7 @@
 #include "private.h"
 #define CRYPTO_LOCK_REF
 #include "sk.h"
+#if !defined(OPENSSL_NO_X509)
 
 #define MYNAME    "x509"
 #define MYVERSION MYNAME " library for " LUA_VERSION " / Nov 2014 / "\
@@ -1181,3 +1182,4 @@ int luaopen_x509(lua_State *L)
 
   return 1;
 }
+#endif
