diff --git a/Makefile.am b/Makefile.am
index 5f640b0..89315ac 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -487,46 +487,6 @@ endif # WITH_PCRE_CPP
 # Each unit test is a binary plus a script that runs that binary in various
 # ways. We install these test binaries in case folks find it helpful.
 
-TESTS += RunTest
-dist_noinst_SCRIPTS += RunTest
-EXTRA_DIST += RunTest.bat
-bin_PROGRAMS += pcretest
-pcretest_SOURCES = pcretest.c
-pcretest_CFLAGS = $(AM_CFLAGS)
-pcretest_LDADD = $(LIBREADLINE)
-if WITH_PCRE8
-pcretest_SOURCES += pcre_printint.c
-pcretest_LDADD += libpcre.la libpcreposix.la
-endif # WITH_PCRE8
-if WITH_PCRE16
-pcretest_SOURCES += pcre16_printint.c
-pcretest_LDADD += libpcre16.la
-endif # WITH_PCRE16
-if WITH_PCRE32
-pcretest_SOURCES += pcre32_printint.c
-pcretest_LDADD += libpcre32.la
-endif # WITH_PCRE32
-if WITH_VALGRIND
-pcretest_CFLAGS += $(VALGRIND_CFLAGS)
-endif # WITH_VALGRIND
-if WITH_GCOV
-pcretest_CFLAGS += $(GCOV_CFLAGS)
-pcretest_LDADD += $(GCOV_LIBS)
-endif # WITH_GCOV
-
-if WITH_PCRE8
-TESTS += RunGrepTest
-dist_noinst_SCRIPTS += RunGrepTest
-bin_PROGRAMS += pcregrep
-pcregrep_SOURCES = pcregrep.c
-pcregrep_CFLAGS = $(AM_CFLAGS)
-pcregrep_LDADD = $(LIBZ) $(LIBBZ2)
-pcregrep_LDADD += libpcre.la libpcreposix.la
-if WITH_GCOV
-pcregrep_CFLAGS += $(GCOV_CFLAGS)
-pcregrep_LDADD += $(GCOV_LIBS)
-endif # WITH_GCOV
-endif # WITH_PCRE8
 
 EXTRA_DIST += \
   testdata/grepbinary \
