diff --git a/src/lua/luvibundle.lua b/src/lua/luvibundle.lua
index ca9ef21..639d7d8 100644
--- a/src/lua/luvibundle.lua
+++ b/src/lua/luvibundle.lua
@@ -332,6 +332,31 @@ local function commonBundle(bundlePaths, mainPath, args)
     mainRequire = require('require')("bundle:main.lua")
   end
 
+  luvi.version = 'logd-custom-build'
+
+  -- Seed Lua's RNG
+  do
+    local math = require('math')
+    local os = require('os')
+    math.randomseed(os.time())
+  end
+
+  -- set global process table
+  local procStat = bundle.stat("deps/process.lua")
+  if procStat and procStat.type == "file" then
+    -- Inject the global process table
+    _G.process = mainRequire('process').globalProcess()
+  end
+
+  -- Load Resolver
+  local dnsStat = bundle.stat("deps/dns.lua")
+  if dnsStat and dnsStat.type == "file" then
+    do
+      local dns = mainRequire('dns')
+      dns.loadResolver()
+    end
+  end
+
   -- Auto-setup global p and libuv version of print
   if mainRequire and (bundle.stat("deps/pretty-print") or bundle.stat("deps/pretty-print.lua")) then
     _G.p = mainRequire('pretty-print').prettyPrint
